### PR TITLE
RATIS-1099. DataStreamServerRpc should connect other peers automatically

### PR DESCRIPTION
**Pull request by szetszwo**
See https://issues.apache.org/jira/browse/RATIS-1099

### CODE REVIEW

File: ratis-netty/src/main/java/org/apache/ratis/netty/server/NettyServerStreamRpc.java [hunk 1]
```diff
@@ -65,23 +68,45 @@
...
+  @Override
+  public void addPeers(Collection<RaftPeer> newPeers) {
+    proxies.addPeers(newPeers);
+    peers.addAll(newPeers);
+  }
+
+  private List<DataStreamOutput> getDataStreamOutput() throws IOException {
+    final List<DataStreamOutput> outs = new ArrayList<>();
+    for(RaftPeer peer : peers) {
+      outs.add(proxies.getProxy(peer.getId()).stream());
```
Comment (amaliujia) [inline ratis-netty/src/main/java/org/apache/ratis/netty/server/NettyServerStreamRpc.java:107]: What if when there is not client of a peer and it tries to create a new one by `newClient`, but still fails to create this client?

**Comment by amaliujia** [comment]
cc @amaliujia

File: ratis-netty/src/main/java/org/apache/ratis/netty/server/NettyServerStreamRpc.java [hunk 2]
```diff
@@ -121,29 +146,30 @@ private void sendReply(DataStreamRequestByteBuf request, ChannelHandlerContext c
   private ChannelInboundHandler getServerHandler(){
     return new ChannelInboundHandlerAdapter(){
       @Override
-      public void channelRead(ChannelHandlerContext ctx, Object msg) {
+      public void channelRead(ChannelHandlerContext ctx, Object msg) throws IOException {
```
Comment (amaliujia) [inline ratis-netty/src/main/java/org/apache/ratis/netty/server/NettyServerStreamRpc.java:149]: Is there a need to add an exception handler in to handle IOException

```
 private ChannelInitializer<SocketChannel> getInitializer(){
    return new ChannelInitializer<SocketChannel>(){
      @Override
      public void initChannel(SocketChannel ch) {
        ChannelPipeline p = ch.pipeline();
        p.addLast(new DataStreamRequestDecoder());
        p.addLast(new DataStreamReplyEncoder());
        p.addLast(getServerHandler());
      }
    };
```?

File: ratis-netty/src/main/java/org/apache/ratis/netty/server/NettyServerStreamRpc.java (hunk 1 above)
Comment (szetszwo) [reply ratis-netty/src/main/java/org/apache/ratis/netty/server/NettyServerStreamRpc.java:107]: peers and proxies must be consistent with each others.  Let me refactor the code to make it clear.

File: ratis-netty/src/main/java/org/apache/ratis/netty/server/NettyServerStreamRpc.java (hunk 2 above)
Comment (szetszwo) [reply ratis-netty/src/main/java/org/apache/ratis/netty/server/NettyServerStreamRpc.java:149]: Let's make streaming working first and ignore exception handling for the moment.  We will work on exception handling in separated issues.  (Just like that we separated the retrying in this issue and didn't want RATIS-1082 becoming to big.)

File: ratis-netty/src/main/java/org/apache/ratis/netty/server/NettyServerStreamRpc.java [hunk 3]
```diff
@@ -121,29 +146,30 @@ private void sendReply(DataStreamRequestByteBuf request, ChannelHandlerContext c
...
+          for (int i = 0; i < outs.size(); i++) {
+            parallelWrites[i + 1] = outs.get(i).getHeaderFuture();
           }
         } else {
           // body
-          for (int i = 0; i < clients.size(); i++) {
-            parallelWrites[i + 1]  =
-              peersStreamOutput.get(request.getStreamId()).get(i).writeAsync(request.slice().nioBuffer());
+          final List<DataStreamOutput> outs = peersStreamOutput.get(request.getStreamId());
+          for (int i = 0; i < peers.size(); i++) {
```
Comment (runzhiwang) [inline ratis-netty/src/main/java/org/apache/ratis/netty/server/NettyServerStreamRpc.java:171]: How about peers.size() -> outs.size() ?

**Comment by runzhiwang** [comment]
@szetszwo LGTM. Just some minor comments and need to fix style error.

File: ratis-netty/src/main/java/org/apache/ratis/netty/server/NettyServerStreamRpc.java (hunk 3 above)
Comment (szetszwo) [reply ratis-netty/src/main/java/org/apache/ratis/netty/server/NettyServerStreamRpc.java:171]: After refactoring, size() is not needed.  Will push a change soon.

File: ratis-netty/src/main/java/org/apache/ratis/netty/server/NettyServerStreamRpc.java (hunk 1 above)
Comment (amaliujia) [reply ratis-netty/src/main/java/org/apache/ratis/netty/server/NettyServerStreamRpc.java]: Thanks for the update. It looks more clear now how does clients and stream output are created and failures are handled.

**Review by amaliujia — APPROVED** [review]
LGTM

**Comment by amaliujia** [comment]
Filed https://issues.apache.org/jira/browse/RATIS-1101 for the flaky UT

**Comment by runzhiwang** [comment]
Reopen pr to trigger CI

**Comment by runzhiwang** [comment]
@szetszwo Thanks the patch. @amaliujia Thanks for review. I have merged it.

**Pull request merged**